feat(token_list): add supply to the token list model

jrainville opened the pull request:
Fixes #11437

Do note that this does **not** perform a migration or a DB update, so for all the tokens already created and added to a community, the description is not updated, so the supply will remain empty.

Since we're still very beta, I think that's acceptable. Updating the description for old tokens is like double the work, so it doesn't seem worth it.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c0044bc3 | [#2](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-nim/job/PR-11685/2/) | 2023-07-26 19:20:22 | ~4 min | `tests/nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-nim/job/PR-11685/2//consoleText) |

caybro commented on src/app/modules/shared_models/token_list_model.nim:
```diff
@@ -76,6 +78,8 @@ QtObject:
       ModelRole.Category.int:"category",
       ModelRole.CommunityId.int:"communityId",
+      ModelRole.Supply.int:"Supply",
+      ModelRole.IsInfiniteSupply.int:"IsInfiniteSupply",
```
```suggestion
      ModelRole.IsInfiniteSupply.int:"isInfiniteSupply",
```

caybro replied:
Role names have to start with a lowercase letter, and we do already refer to them throughout the code as e.g. `model.supply`

caybro replied:
This should even read `infiniteSupply` as that's what all the UI elements use currently:

![image](https://github.com/status-im/status-desktop/assets/5377645/9c1174ad-b524-43c9-ba7e-0251c70f7e8b)

caybro commented on src/app/modules/shared_models/token_list_model.nim:
```diff
@@ -76,6 +78,8 @@ QtObject:
       ModelRole.Image.int:"icon",
       ModelRole.Category.int:"category",
       ModelRole.CommunityId.int:"communityId",
+      ModelRole.Supply.int:"Supply",
```
```suggestion
      ModelRole.Supply.int:"supply",
```

caybro commented on src/app/modules/shared_models/token_list_item.nim:
```diff
@@ -14,6 +14,8 @@ type
     image*: string
     category*: int
     communityId*: string
+    supply*: string
```
Should this really be a `string`?

micieslak replied:
Imo string is ok here, especially if we want to keep supply (not sure if now) expressed in base units where int or float may be insufficient.

alexjba reviewed (approved):
LGTM! Other than the issue pointed out by Lukas.

jrainville commented:
Re-requesting reviews because I changed the code a little. It uses ` getCommunityTokens` instead. I dropped the status-go changes as well. See the updated description for the details

caybro commented on src/app/modules/shared_models/token_list_model.nim:
```diff
@@ -92,7 +96,7 @@ QtObject:
         result = newQVariant(item.getKey())
       of ModelRole.Name:
-        result = newQVariant(item.getName())
+        result = newQVariant(item.getName() & " - " & item.getSupply() & " - " & $item.getInfiniteSupply())
```
Why? :)

caybro reviewed (commented):
Otherwise LGTM

alexjba commented:
> I updated the code so that it uses getCommunityTokens instead of the CommunityTokenMetadata like it used to.

I've had a similar attempt some time ago when analysing #10546, but I've got stuck by two issues:
- `getCommunityTokens` will return the tokens minted by the current account. Currently these tokens are stored in a single DB. So it won't work for an owner with multiple devices - we're probably missing the tokens sync feature.
- AFAIK `getCommunityTokens` will not work for admins as they don't have the rights to mint/airdrop etc. But they have the rights to create/edit permissions and they need the supply amount.

caybro reviewed (approved):
LGTM

0x-r4bbit commented:
> getCommunityTokens will return the tokens minted by the current account. Currently these tokens are stored in a single DB. So it won't work for an owner with multiple devices - we're probably missing the tokens sync feature.

@alexjba This is being done here https://github.com/status-im/status-go/pull/3730 

> AFAIK getCommunityTokens will not work for admins as they don't have the rights to mint/airdrop etc. But they have the rights to create/edit permissions and they need the supply amount.

That is correct. Admins and other members only have information about community tokens derived from the community token metadata atm.

However, I think there are plans to share the locally saved community token data among admins/tokenmasters/owners

0x-r4bbit commented:
@0x-r4bbit is there any way that you could extend the work you're doing here https://github.com/status-im/status-go/pull/3730 to be done for admins and TokenMasters too?

I can look into it but that'd be a separate effort as I think we'd have to introduce a new signal for that.
We might need that signal for other kinds of data that wants to be shared among admins.

However, atm I'm still busy with removing revealed addresses from members and that requires some bigger changes.
I'll create an issue so we have an eye on it.

jrainville commented:
> I can look into it but that'd be a separate effort as I think we'd have to introduce a new signal for that.

No problem, it can be another issue and PR. I was curious if it was only a matter of changing a small thing to make it propagate to admins also.

> I'll create an issue so we have an eye on it.

Greta thanks

0x-r4bbit commented:
@jrainville added an issue here with an initial brain dump, feel free to add more considerations and discuss details there: https://github.com/status-im/status-desktop/issues/11715

jrainville commented:
@alexjba @0x-r4bbit I added a fallback. So it will use the tokenMetadata all the time and try to get the supply from the CommunityTokens. If we don't have the CommunityTokens (for example an admin before we add the propagation of tokens, or just an admin that missed the last propagation), we'll still be able to show the tokens, but the supply with be infinite.

jrainville commented:
FYI for QAs, there isn't much to test. The supply doesn't appear in the UI yet. You can test that the token lists in all the places still work I guess

lukaszso commented:
@jrainville could you rebase this one plz? We need [this fix](https://github.com/status-im/status-desktop/issues/11762) in order to be able to test.

jrainville commented:
> @jrainville could you rebase this one plz? We need [this fix](https://github.com/status-im/status-desktop/issues/11762) in order to be able to test.

@lukaszso done

lukaszso reviewed (approved):
Looks good :)
